### PR TITLE
QDQ tool modification part2

### PR DESCRIPTION
**Pull request by chilo-ms**
This PR added three features:

1. The default behavior is inserting a QDQ pair to one tensor and multiple nodes can share a QDQ pair as their inputs if needed. 
    In TRT, QDQ pair can’t be shared between nodes, so it will create dedicated QDQ pairs for each node.
    Add a new extra options DedicatedQDQPair = True/False : Default is False. When inserting QDQ pair, multiple nodes can share a single QDQ pair as their inputs. If True, it will create identical and dedicated QDQ pair for each node.

2. Add a new extra options OpTypesSupportPerChannelQuantization = list of op type : Default is []. List of op types that has per channel quantization support

3. Add a new extra options QDQChannelAxis = Integer. Default is 0. Channel axis for QDQ pair when per_channel is True.

### CODE REVIEW

**Comment by chilo-ms** [comment]
I'm adding unit tests and will add them later.

**Comment by chilo-ms** [comment]
> I'm adding unit tests and will add them later.

unit tests added

File: onnxruntime/python/tools/quantization/operators/qdq_base_operator.py [hunk 1]
```diff
@@ -19,4 +19,7 @@ def quantize(self):
             nodes_to_iterate = itertools.chain(node.input, node.output)
 
         for tensor_name in nodes_to_iterate:
-            self.quantizer.quantize_tensor(tensor_name)
+            if self.quantizer.is_per_channel():
+                self.quantizer.quantize_tensor_per_channel(tensor_name, self.quantizer.qdq_channel_axis)
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/operators/qdq_base_operator.py:23]: what if the operator doesn't support per-channel

Comment (yufenglee) [reply onnxruntime/python/tools/quantization/operators/qdq_base_operator.py:23]: cpu ep only per-channel for Conv and Matmul now

File: onnxruntime/python/tools/quantization/quantize.py [hunk 2]
```diff
@@ -194,6 +194,11 @@ def quantize_static(model_input,
                                               inserts both QuantizeLinear/DeQuantizeLinear nodes to weight.
             OpTypesToExcludeOutputQuantizatioin = list of op type : Default is []. If any op type is specified, it won't quantize  
                                                                     the output of ops with this specific op types.
+            DedicatedQDQPair = True/False : Default is False. When inserting QDQ pair, multiple nodes can share a single QDQ pair as their inputs.
+                                            If True, it will create identical and dedicated QDQ pair for each node.  
+            AddQDQToAddNodeFollowedByReduceMeanNode = True/False : Default is False. It adds QDQ pairs to every Add node if Add op type is in op_types_to_quantize.
+                                                                   If True, only Add node followed by ReduceMean node is going to be added QDQ pair.
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/quantize.py:200]: This is too specific.

Comment (yufenglee) [reply onnxruntime/python/tools/quantization/quantize.py:200]: we have a general way nodes_to_exclude. why not use it

Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/quantize.py:200]: Actually I'm using nodes_to_exclude in qdq_quantizer.py#128.
Agree it's too specific. I'm removing this feature and move the logic to e2e example to get nodes_to_exclude.

File: onnxruntime/python/tools/quantization/operators/qdq_base_operator.py (hunk 1 above)
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/operators/qdq_base_operator.py]: Add a new extra options OpTypesSupportPerChannelQuantization

File: onnxruntime/python/tools/quantization/quantize.py [hunk 3]
```diff
@@ -194,6 +194,10 @@ def quantize_static(model_input,
                                               inserts both QuantizeLinear/DeQuantizeLinear nodes to weight.
             OpTypesToExcludeOutputQuantizatioin = list of op type : Default is []. If any op type is specified, it won't quantize  
                                                                     the output of ops with this specific op types.
+            DedicatedQDQPair = True/False : Default is False. When inserting QDQ pair, multiple nodes can share a single QDQ pair as their inputs.
+                                            If True, it will create identical and dedicated QDQ pair for each node.  
+            OpTypesSupportPerChannelQuantization = list of op type : Default is []. List of op types that has per channel quantization support.
+            QDQChannelAxis = Integer : Default is 0. Channel axis for QDQ pair when per_channel is True.
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/quantize.py:200]: >QDQChannelAxis [](http://example.com/codeflow?start=12&length=14)

i don't think all the operators can share same per-channel axis.

Comment (yufenglee) [reply onnxruntime/python/tools/quantization/quantize.py:200]: which operators in TRT supports per-channel?

File: onnxruntime/python/tools/quantization/quantize.py [hunk 4]
```diff
@@ -194,6 +194,10 @@ def quantize_static(model_input,
                                               inserts both QuantizeLinear/DeQuantizeLinear nodes to weight.
             OpTypesToExcludeOutputQuantizatioin = list of op type : Default is []. If any op type is specified, it won't quantize  
                                                                     the output of ops with this specific op types.
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/quantize.py:196]: as TRT has so many , i'm thinking we may need a config file for different execution provider

File: onnxruntime/python/tools/quantization/quantize.py (hunk 3 above)
Comment (stevenlix) [reply onnxruntime/python/tools/quantization/quantize.py:200]: In TRT pretty much we only use per-channel for weights. All activations are using per-tensor.

**Review by stevenlix — APPROVED** [review]

**Comment by stevenlix** [comment]
Talked to @yufenglee offline. Keep flags OpTypesSupportPerChannelQuantization and QDQChannelAxis for TRT internal use for now.  Will have a followup PR to fine tune the related code.

File: onnxruntime/python/tools/quantization/operators/qdq_base_operator.py [hunk 5]
```diff
@@ -19,4 +20,13 @@ def quantize(self):
...
         for tensor_name in nodes_to_iterate:
-            self.quantizer.quantize_tensor(tensor_name)
+            if self.quantizer.is_per_channel():
+                if node.op_type in self.quantizer.op_types_support_per_channel_quantization :
+                    self.quantizer.quantize_tensor_per_channel(tensor_name, self.quantizer.qdq_channel_axis)
+                else:
+                    logging.warning(
+                        "{} doesn't support per channel quantization. Quantize tensor: {} with per-tensor instead.".format(
+                            node.op_type, tensor_name))
+                    self.quantizer.quantize_tensor(tensor_name)
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/operators/qdq_base_operator.py:30]: it should not be a warning.

**Review by yufenglee — APPROVED** [review]

**Pull request merged**